// typeorm-extension/dist/connection/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildConnectionOptions = void 0;
var typeorm_1 = require("typeorm");
var index_1 = require("../index");
var utils_1 = require("./utils");
__exportStar(require("./type"), exports);
__exportStar(require("./utils"), exports);
function buildConnectionOptions(options) {
    var _a, _b, _c;
    return __awaiter(this, void 0, void 0, function () {
        var connectionOptionsReader, connectionOptions, distDirectory, tsCompilerOptions, e_1;
        return __generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    options = options !== null && options !== void 0 ? options : {};
                    connectionOptionsReader = new typeorm_1.ConnectionOptionsReader({
                        root: (_a = options.root) !== null && _a !== void 0 ? _a : process.cwd(),
                        configName: (_b = options.configName) !== null && _b !== void 0 ? _b : 'ormconfig',
                    });
                    return [4 /*yield*/, connectionOptionsReader.get((_c = options.name) !== null && _c !== void 0 ? _c : 'default')];
                case 1:
                    connectionOptions = _d.sent();
                    /* istanbul ignore next */
                    if (options.buildForCommand) {
                        Object.assign(connectionOptions, {
                            subscribers: [],
                            synchronize: false,
                            migrationsRun: false,
                            dropSchema: false,
                            logging: ["query", "error", "schema"]
                        });
                    }
                    connectionOptions = index_1.createDefaultSeederOptions(connectionOptions);
                    _d.label = 2;
                case 2:
                    _d.trys.push([2, 4, , 5]);
                    return [4 /*yield*/, index_1.getCompilerOptions(options.tsConfigDirectory)];
                case 3:
                    tsCompilerOptions = _d.sent();
                    distDirectory = tsCompilerOptions.outDir;
                    return [3 /*break*/, 5];
                case 4:
                    e_1 = _d.sent();
                    return [3 /*break*/, 5];
                case 5:
                    utils_1.modifyConnectionOptionsForRuntimeEnvironment(connectionOptions, {
                        distDirectory: distDirectory
                    });
                    return [2 /*return*/, connectionOptions];
            }
        });
    });
}
exports.buildConnectionOptions = buildConnectionOptions;
